### PR TITLE
Implement DLR LibUtils

### PR DESCRIPTION
**Pull request by stu1130**
## Channel for questions ##
Before or while filing an issue please feel free to join our [Slack channel](https://join.slack.com/t/deepjavalibrary/shared_invite/zt-ar91gjkz-qbXhr1l~LFGEIEeGBibT7w) to get in touch with development team, ask questions, find out what's cooking and more!

## Description ##
(Brief description of what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- [ ] Code is well-documented: 
    - For user-facing API changes, Java doc has been updated. 
    - For new examples, README.md is added to explain the what the example does.
- [ ] To the my best knowledge, [examples](https://github.com/awslabs/djl/tree/master/examples) and [jupyter notebooks](https://github.com/awslabs/djl/tree/master/jupyter) are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, Java doc)
- [ ] Feature2, tests, (and when applicable, Java doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made?
- Interesting edge cases to note here

### CODE REVIEW

File: dlr/dlr-engine/build.gradle [hunk 1]
```diff
@@ -8,14 +8,45 @@ repositories {
...
     testImplementation "org.slf4j:slf4j-simple:${slf4j_version}"
     testRuntimeOnly project(":pytorch:pytorch-engine")
+    testRuntimeOnly "ai.djl.dlr:dlr-native-auto:${dlr_version}-SNAPSHOT"
     testRuntimeOnly "ai.djl.pytorch:pytorch-native-auto:${pytorch_version}"
 }
 
+processResources {
+    outputs.dir file("${project.buildDir}/jnilib")
+    doLast {
+        def url = "https://djl-ai.s3.amazonaws.com/publish/dlr-${dlr_version}/jnilib/${djl_version}"
```
Comment (frankfliu) [inline dlr/dlr-engine/build.gradle:24]: https://publish.djl.ai/

File: dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/LibUtils.java [hunk 2]
```diff
@@ -24,11 +44,302 @@
...
+            Pattern.compile("(\\d+\\.\\d+\\.\\d+(-\\w)?)(-SNAPSHOT)?(-\\d+)?");
+
     private LibUtils() {}
 
     public static void loadLibrary() {
-        // TODO implement
-        System.load("/Users/leecheng/workspace/djl/dlr/dlr-native/libdlr.dylib");
-        System.load("/Users/leecheng/.djl.ai/dlr/1.5.0-cpu-osx-x86_64/libdjl_dlr.dylib");
+        String nativePath = findNativeOverrideLibrary();
+        String jniPath;
```
Comment (frankfliu) [inline dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/LibUtils.java:59]: To make is simple, we should always copy jni to our cachDir

Comment (stu1130) [reply dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/LibUtils.java]: done

File: dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/LibUtils.java [hunk 3]
```diff
@@ -24,11 +44,302 @@
...
+                throw new IllegalStateException(
+                        "Failed to read DLR native library jar properties", e);
+            }
+            if (platform == null) {
+                throw new IllegalStateException(
+                        "Your DLR native library jar does not match your operating system. Make sure the Maven Dependency Classifier matches your system type.");
+            }
+            cacheDir = getCacheDir(platform);
+        }
+        String jniPath = copyJniLibraryFromClasspath(cacheDir, fallback.get());
```
Comment (frankfliu) [inline dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/LibUtils.java:83]: DLR jni should be work for both CPU/GPU build right? then we don't need fallback.

Comment (stu1130) [reply dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/LibUtils.java]: right! removed it

File: dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/LibUtils.java [hunk 4]
```diff
@@ -24,12 +44,302 @@
...
     private LibUtils() {}
 
     public static void loadLibrary() {
-        // TODO implement
-        System.load("/Users/leecheng/.djl.ai/dlr/1.5.0-SNAPSHOT-cpu-osx-x86_64/libdlr.dylib");
-        System.load(
-                "/Users/leecheng/.djl.ai/dlr/1.5.0-SNAPSHOT-cpu-osx-x86_64/0.9.0-SNAPSHOT-cpu-libdjl_dlr.dylib");
+        String nativePath = findNativeOverrideLibrary();
+        Path cacheDir;
+        AtomicBoolean fallback = new AtomicBoolean(false);
```
Comment (frankfliu) [inline dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/LibUtils.java:60]: This variable is longer needed.

Comment (stu1130) [reply dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/LibUtils.java:60]: The variable is for findNativeLibrary

**Review by frankfliu — APPROVED** [review]

**Pull request merged**